File: apps/macos/RealEstateIntel/Models/Analysis.swift
import Foundation

// MARK: - Top-Level Analysis Response

struct ListingAnalysis: Codable, Sendable {
    let listingId: Int
    let summary: AnalysisSummary
    let locationContext: AnalysisLocationContext
    let buildingContext: AnalysisBuildingContext?
    let marketSaleContext: AnalysisMarketContext?
    let marketRentContext: AnalysisMarketRentEstimate?
    let investorMetrics: AnalysisInvestorMetrics?
    let riskFlags: [String]
    let upsideFlags: [String]
    let assumptions: [String]
    let missingData: [String]
    let legalRentSummary: AnalysisLegalRentSummary?
    let confidence: AnalysisConfidence
    let computedAt: String
}

// MARK: - Summary

struct AnalysisSummary: Codable, Sendable {
    let headline: String?
    let keyFacts: [String]
}

// MARK: - Location Context

struct AnalysisLocationContext: Codable, Sendable {
    let districtNo: Int?
    let districtName: String?
    let nearestTransit: String?
    let nearestTransitDistanceM: Double?
    let parksNearby: Int
    let schoolsNearby: Int
}

// MARK: - Building Context

struct AnalysisBuildingContext: Codable, Sendable {
    let buildingFactId: Int
    let matchConfidence: String
    let yearBuilt: Int?
    let typology: String?
    let unitCount: Int?
    let source: String
    let sourceUpdatedAt: String?
}

// MARK: - Market Sale Context

struct AnalysisMarketContext: Codable, Sendable {
    let fallbackLevel: String
    let sampleSize: Int
    let medianPpsqm: Int?
    let p25Ppsqm: Int?
    let p75Ppsqm: Int?
    let confidence: String
}

// MARK: - Market Rent Estimate

struct AnalysisMarketRentEstimate: Codable, Sendable {
    let estimateLow: Double?
    let estimateMid: Double?
    let estimateHigh: Double?
    let eurPerSqmMid: Double?
    let fallbackLevel: String
    let sampleSize: Int
    let confidence: String
}

// MARK: - Investor Metrics

struct AnalysisInvestorMetrics: Codable, Sendable {
    let grossYield: GrossYield
    let priceToRent: Double?
    let sensitivityBands: SensitivityBands

    struct GrossYield: Codable, Sendable {
        let value: Double?
        let assumptions: [String]
    }

    struct SensitivityBands: Codable, Sendable {
        let low: Double?
        let base: Double?
        let high: Double?
    }
}

// MARK: - Legal-Rent Summary

struct AnalysisLegalRentSummary: Codable, Sendable {
    let status: String
    let regimeCandidate: String?
    let confidence: String
    let strongSignals: [LegalRentSignal]
    let weakSignals: [LegalRentSignal]
    let missingFacts: [String]
    let reviewRequired: Bool
    let indicativeBandLow: Double?
    let indicativeBandHigh: Double?
    let disclaimer: String

    struct LegalRentSignal: Codable, Sendable {
        let signal: String
        let source: String
    }
}

// MARK: - Confidence

struct AnalysisConfidence: Codable, Sendable {
    let level: String
    let degradationReasons: [String]
}

File: packages/contracts/src/analysis.ts
// ── Comparable Listing ─────────────────────────────────────────────────────

export interface ComparableEntry {
  listingId: number;
  title: string;
  districtNo: number | null;
  operationType: string;
  propertyType: string;
  listPriceEurCents: number | null;
  pricePerSqmEur: number | null;
  livingAreaSqm: number | null;
  rooms: number | null;
  distanceM: number | null;
  firstSeenAt: Date;
  canonicalUrl: string;
  /** Why this comparable was selected (e.g., "nearby similar apartment") */
  matchReason?: string;
  /** Days since the comparable was first seen */
  recencyDays?: number;
  /** How close the area is to the target listing (percentage, 100 = identical) */
  areaSimilarityPct?: number | null;
  /** Difference in room count vs target listing */
  roomDiff?: number | null;
}

export type ComparableFallbackLevel = 'nearby' | 'district' | 'city';
export type ConfidenceLevel = 'high' | 'medium' | 'low';

export interface MarketContext {
  comparables: ComparableEntry[];
  fallbackLevel: ComparableFallbackLevel;
  sampleSize: number;
  medianPpsqm: number | null;
  p25Ppsqm: number | null;
  p75Ppsqm: number | null;
  confidence: ConfidenceLevel;
}

// ── Market Rent Estimate ──────────────────────────────────────────────────

export interface MarketRentEstimate {
  estimateLow: number | null;
  estimateMid: number | null;
  estimateHigh: number | null;
  eurPerSqmMid: number | null;
  fallbackLevel: ComparableFallbackLevel;
  sampleSize: number;
  confidence: ConfidenceLevel;
}

// ── Investor Metrics ──────────────────────────────────────────────────────

export interface InvestorMetrics {
  grossYield: {
    value: number | null;
    assumptions: string[];
  };
  priceToRent: number | null;
  sensitivityBands: {
    low: number | null;
    base: number | null;
    high: number | null;
  };
}

// ── Building Context ──────────────────────────────────────────────────────

export interface BuildingContext {
  buildingFactId: number;
  matchConfidence: string;
  yearBuilt: number | null;
  typology: string | null;
  unitCount: number | null;
  source: string;
  sourceUpdatedAt: Date | null;
}

// ── Legal-Rent Summary ──────────────────────────────────────────────────

export interface LegalRentSummary {
  status: string;
  regimeCandidate: string | null;
  confidence: string;
  strongSignals: Array<{ signal: string; source: string }>;
  weakSignals: Array<{ signal: string; source: string }>;
  missingFacts: string[];
  reviewRequired: boolean;
  indicativeBandLow: number | null;
  indicativeBandHigh: number | null;
  disclaimer: string;
}

// ── Analysis Confidence ──────────────────────────────────────────────────

export interface AnalysisConfidence {
  level: ConfidenceLevel;
  degradationReasons: string[];
}

// ── Listing Analysis ──────────────────────────────────────────────────────

export interface ListingAnalysis {
  listingId: number;
  summary: {
    headline: string;
    keyFacts: string[];
  };
  locationContext: {
    districtNo: number | null;
    districtName: string | null;
    nearestTransit: string | null;
    nearestTransitDistanceM: number | null;
    parksNearby: number;
    schoolsNearby: number;
  };
  buildingContext: BuildingContext | null;
  marketSaleContext: MarketContext | null;
  marketRentContext: MarketRentEstimate | null;
  investorMetrics: InvestorMetrics | null;
  riskFlags: string[];
  upsideFlags: string[];
  assumptions: string[];
  missingData: string[];
  legalRentSummary: LegalRentSummary | null;
  confidence: AnalysisConfidence;
  computedAt: Date;
}
